### ISSUE TITLE
[Bug]: apply_temperature may cause nan in probs

### ISSUE THREAD

**Issue by courage17340** [vllm-project/vllm#22180]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

Recently, we met some illegal memory access problems. When we turned off flashinfer sampling, the problem disappeared, or at least it was much harder to trigger. Unfortunately, I can't provide a minimal reproduction here.

After some investigation, I believe the problem is caused by `apply_temperature`. Suppose we have a decode batch with both random sampling and greedy sampling. [Here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/worker/gpu_input_batch.py#L306) will set the temperature for greedy sampling to -1.0, and [here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/sample/sampler.py#L99) will div all the logits against temperature. If the logits (for greedy sampling) contains `-inf`, it will become `inf` after the division, and hence the probs tensor will be fully `nan`, I guess maybe it will cause bad things when using flashinfer top_p sampling. I'm not 100% sure about this, because I use probs with `nan` to run against flashinfer top_p sampling, and compute-sanitizer can't find any problems. But anyway, `nan` in probs is dangerous, I think it should be fixed.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by youkaichao** [comment]
cc @njhill 
I think the divide by -1.0 is introduced by https://github.com/vllm-project/vllm/pull/13587

**Comment by njhill** [comment]
Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...

@courage17340 do you mean it impacts all rows of the logits tensor not just those whose logits contain `inf`?

**Comment by youkaichao** [comment]
> Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...

we do have lots of reports about IMA from sampling, and we don't know what's the root cause yet. this issue description looks reasonable.

What's the original motivation for dividing by -1.0 ? it seems to be incorrect from my understanding.

**Comment by courage17340** [comment]
> Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...
> 
> [@courage17340](https://github.com/courage17340) do you mean it impacts all rows of the logits tensor not just those whose logits contain `inf`?

It looks like all of the conditions are required to reproduce:
1. the batch should contain both greedy sampling (for temperature=-1) and random sampling (otherwise all_greedy is satisfied and apply_temperature is not executed)
2. the original logits for greedy sampling contains `-inf`, normally set by some logits modification (bad words or others?)
3. flashinfer sampling is used, which means the batch contains top_p/top_k sampling

The whole batch will then be sampled by flashinfer top_p/top_k sampling kernel, and the rows with `inf` in logits will get `nan` in probs, and other rows looks OK. I don't know what happens next because compute-sanitizer can't find errors when I just run the kernel on the dumped tensors directly.

**Comment by njhill** [comment]
But these rows are the ones that were already separately greedy sampled and the corresponding logits in this method are discarded (filtered out by the subsequent `where` [here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/sample/sampler.py#L145-L149)). Do you have a test case where `nan` values are actually returned to the user?

**Comment by courage17340** [comment]
> But these rows are the ones that were already separately greedy sampled and the corresponding logits in this method are discarded (filtered out by the subsequent `where` [here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/sample/sampler.py#L145-L149)). Do you have a test case where `nan` values are actually returned to the user?

I guess that `nan` in probs may trigger some corner cases and some kernel (like flashinfer top_p sampling) may access invalid indices. However, I still can't verify that yet.

**Comment by zifeitong** [comment]
Thanks @courage17340, I think your theory is correct. 

@njhill I think the key issue is `torch.softmax()` called by [flashinfer](https://github.com/vllm-project/vllm/blob/4b3e4474d73ae9cf0d6c8315570fdffd71037d08/vllm/v1/sample/ops/topk_topp_sampler.py#L289) will return `nan` for the entire row if there are any `inf` in the row, which cause flahsinfer crashing. The filter is too late at this moment.

I tested a hack by clamping logits to [float.MIN, float.MAX] on our workload and it fixes the crashes in #19483. The proper fix should be avoiding creating `inf` in the first place.

**Comment by zifeitong** [comment]
> Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...

I think it's not that easy to trigger this bug, you'll need both greedy sampling (which is never default setting) and sampling params causing "-inf" (bad words, structured output) mixed in a single batch.

**Comment by ashgold** [comment]
> Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...
> 
> [@courage17340](https://github.com/courage17340) do you mean it impacts all rows of the logits tensor not just those whose logits contain `inf`?

It was said that it was six months ago, so I doubt that this is the cause.
This problem also occurred at v0.8.4.
It does not occur at 0.7.2.
This problem occurs very occasionally only in operational situations, so it is very difficult to reproduce, so it does not seem to have become an issue so far.

**Comment by njhill** [comment]
> > Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...
> 
> I think it's not that easy to trigger this bug, you'll need both greedy sampling (which is never default setting) and sampling params causing "-inf" (bad words, structured output) mixed in a single batch.

@courage17340 @zifeitong We should be able to create a reproducer with all of these things though? (i.e. which also uses flashifer sampling)

**Comment by courage17340** [comment]
> > > Hmm, that change was made 6 months ago so it's a bit surprising the problem hasn't shown up earlier if that's the cause...
> > 
> > 
> > I think it's not that easy to trigger this bug, you'll need both greedy sampling (which is never default setting) and sampling params causing "-inf" (bad words, structured output) mixed in a single batch.
> 
> [@courage17340](https://github.com/courage17340) [@zifeitong](https://github.com/zifeitong) We should be able to create a reproducer with all of these things though? (i.e. which also uses flashifer sampling)

I have a very complex reproduction, but I can't make a minimal reproduction and I don't know why. I dumped the tensors (probs & top_p) and even rng states before the illegal memory access happen. But when I called flashinfer sampling with the tensors, nothing happened, and compute-sanitizer couldn't find any errors too. That's why I said I'm not 100% sure about this.

**Comment by rishitdholakia13** [comment]
I root caused a similar issue I was facing with v0.10, I have a fix up for it : I root caused the issue with the v0.10 version and and have a fix up : https://github.com/vllm-project/vllm/pull/22963. Hopefully it helps answer the reason for the illegal memory access during topp sampling using flashinfer

**Comment by dskhudia** [comment]
The root cause of https://github.com/flashinfer-ai/flashinfer/issues/1575 is also I think inf/nan in logits/probs.

**Comment by courage17340** [comment]
@zifeitong  Hello, do you have any updates for your PR?
